### ISSUE TITLE
WGBS: unrecognized reference name ""; treated as unmapped

### ISSUE THREAD

**Issue by hanrong498** [maxplanck-ie/snakepipes#936]
Hi!

Sorry for the spamming :)

I am trying to run the WGBS pipeline but got the error in`bwameth/logs/211029_1_FZ_WGBS_HOIK_p35_rep1_S1_L001.map_reads.err `:

```
Traceback (most recent call last):
  File "/work/upzenk/miniconda3/envs/20d3d6a39823b295d3503a5332b4207b/bin/bwameth.py", line 559, in <module>
    main(sys.argv[1:])
  File "/work/upzenk/miniconda3/envs/20d3d6a39823b295d3503a5332b4207b/bin/bwameth.py", line 550, in main
    bwa_mem(args.reference, conv_fqs_cmd, ' '.join(map(str, pass_through_args)),
  File "/work/upzenk/miniconda3/envs/20d3d6a39823b295d3503a5332b4207b/bin/bwameth.py", line 349, in bwa_mem
    raise BWAMethException("first run bwameth.py index %s OR bwameth.py index-mem2 %s" % (fa, fa))
BWAMethException: first run bwameth.py index  OR bwameth.py index-mem2 
[W::hts_set_opt] Cannot change block size for this format
samtools sort: failed to read header from "-"
```

The shell command from `WGBS_run` is 

```
shell:
        
            TMPDIR=/tmp/
            MYTEMP=$(mktemp -d "${TMPDIR:-/tmp}"/snakepipes.XXXXXXXXXX)
            bwameth.py --threads 12 --reference "" "FASTQ/211029_1_FZ_WGBS_HOIK_p35_rep1_S1_L001_R1_001_MM_1.fastq.gz" "FASTQ/211029_1_FZ_WGBS_HOIK_p35_rep1_S1_L001_R2_001_MM_1.fastq.gz" 2> bwameth/logs/211029_1_FZ_WGBS_HOIK_p35_rep1_S1_L001.map_reads.err | 	        samtools sort -T "$MYTEMP"/211029_1_FZ_WGBS_HOIK_p35_rep1_S1_L001 -m 3G -@ 4 -o "bwameth/211029_1_FZ_WGBS_HOIK_p35_rep1_S1_L001.bam" 2>> bwameth/logs/211029_1_FZ_WGBS_HOIK_p35_rep1_S1_L001.map_reads.err
            rm -rf "$MYTEMP"
```

when I activate this env I have the samtools:
```
(20d3d6a39823b295d3503a5332b4207b) [hhu@jed ~]$ samtools --help

Program: samtools (Tools for alignments in the SAM format)
Version: 1.17 (using htslib 1.17)
```

Can you help me with that? Thanks!

**Comment by hanrong498** [comment]
I think the problem is that the reference.fa is missing. I am trying to manually re-write the organism.yaml file

```
$ cat /work/upzenk/miniconda3/envs/snakePipes/lib/python3.11/site-packages/snakePipes/shared/organisms/GRCh38_v39_lambda_spikein.yaml 
blacklist_bed: ''
bowtie2_index: /scratch/hhu/index_snakepipes/human_GRCh38_v39_lambda_spikein/BowtieIndex/genome
bwa_index: /scratch/hhu/index_snakepipes/human_GRCh38_v39_lambda_spikein/BWAIndex/genome.fa
bwa_mem2_index: /scratch/hhu/index_snakepipes/human_GRCh38_v39_lambda_spikein/BWA-MEM2Index/genome.fa
bwameth2_index: /scratch/hhu/index_snakepipes/human_GRCh38_v39_lambda_spikein/BWAmeth2Index/genome.fa
bwameth_index: /scratch/hhu/index_snakepipes/human_GRCh38_v39_lambda_spikein/BWAmethIndex/genome.fa
```

All the indexes were generated using the `createIndices` pipeline. I have a naive question: among all the files in the `*Index/` folder, do we always put the `/genome.fa` here? And if I understand it correctly, /genome.fa here is actually a symlink to `genome_fasta/genome.fa`?

**Comment by hanrong498** [comment]
Hi, 

Here I fixed the reference genome issue but still have problems running the WGBS pipeline. In the `.err`, the bwameth was run multiple times and the mapped files seem to not have a reference name:
```
[M::process] 0 single-end sequences; 1818182 paired-end sequences
[M::process] read 1818182 sequences (120000012 bp)...
[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (36, 178823, 11466, 10)
[M::mem_pestat] analyzing insert size distribution for orientation FF...
[M::mem_pestat] (25, 50, 75) percentile: (117, 362, 1187)
[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 3327)
[M::mem_pestat] mean and std.dev: (635.83, 645.35)
[M::mem_pestat] low and high boundaries for proper pairs: (1, 4397)
[M::mem_pestat] analyzing insert size distribution for orientation FR...
[M::mem_pestat] (25, 50, 75) percentile: (83, 119, 169)
[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 341)
[M::mem_pestat] mean and std.dev: (126.55, 60.59)
[M::mem_pestat] low and high boundaries for proper pairs: (1, 427)
[M::mem_pestat] analyzing insert size distribution for orientation RF...
[M::mem_pestat] (25, 50, 75) percentile: (113, 314, 722)
[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1940)
[M::mem_pestat] mean and std.dev: (437.26, 432.76)
[M::mem_pestat] low and high boundaries for proper pairs: (1, 2549)
[M::mem_pestat] analyzing insert size distribution for orientation RR...
[M::mem_pestat] (25, 50, 75) percentile: (119, 633, 788)
[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2126)
[M::mem_pestat] mean and std.dev: (557.10, 407.29)
[M::mem_pestat] low and high boundaries for proper pairs: (1, 2795)
[M::mem_pestat] skip orientation FF
[M::mem_pestat] skip orientation RR
[M::mem_process_seqs] Processed 1818182 reads in 5503.742 CPU sec, 773.964 real sec
[M::process] 0 single-end sequences; 1818182 paired-end sequences
[M::process] read 1818182 sequences (120000012 bp)...
[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (33, 178223, 11445, 7)
[M::mem_pestat] analyzing insert size distribution for orientation FF...
[M::mem_pestat] (25, 50, 75) percentile: (109, 216, 777)
[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2113)
[M::mem_pestat] mean and std.dev: (373.16, 372.35)
[M::mem_pestat] low and high boundaries for proper pairs: (1, 2781)
[M::mem_pestat] analyzing insert size distribution for orientation FR...
[M::mem_pestat] (25, 50, 75) percentile: (83, 119, 169)
[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 341)
[M::mem_pestat] mean and std.dev: (126.27, 60.50)
[M::mem_pestat] low and high boundaries for proper pairs: (1, 427)
[M::mem_pestat] analyzing insert size distribution for orientation RF...
[M::mem_pestat] (25, 50, 75) percentile: (109, 309, 722)
[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1948)
[M::mem_pestat] mean and std.dev: (437.98, 439.44)
[M::mem_pestat] low and high boundaries for proper pairs: (1, 2561)
[M::mem_pestat] skip orientation RR as there are not enough pairs
[M::mem_pestat] skip orientation FF
rse1] unrecognized reference name ""; treated as unmapped
[W::sam_parse1] unrecognized reference name ""; treated as unmapped
[W::sam_parse1] unrecognized reference name ""; treated as unmapped
[W::sam_parse1] unrecognized reference name ""; treated as unmapped
[W::sam_parse1] unrecognized reference name ""; treated as unmapped
[W::sam_parse1] unrecognized reference name ""; treated as unmapped
```
This is not the complete log. The bwamem was executed > 10 times and there are more `[W::sam_parse1] unrecognized reference name ""; treated as unmapped` message below.

Can you help me with it? Thanks!

**Comment by hanrong498** [comment]
This issue of `[W::sam_parse1] unrecognized reference name ""; treated as unmapped` from `bwameth.py` was mentioned [here](https://www.biostars.org/p/9552502/)

As I have just checked, the reference I generated from bwa was also `rchr1 1`, which probably makes it unrecognisable.
```
$ head genome.fa.bwameth.c2t
>rchr1 1
```

Same issue was mentioned previously #898 

Maybe there is a way to work around it?

**Comment by katsikora** [comment]
Hi @hanrong498,

thanks for looking into the details of the issue. Indeed, I've seen #898, and I wasn't able to reproduce it here.
Unless something got messed up with bwa-meth or another tool on conda since we've built and tested the release. 
Did you build your conda environments for snakePipes recently?
I can try if I can reproduce it with a freshly made conda env.

Best wishes,

Katarzyna

**Comment by hanrong498** [comment]
Hi Katarzyna,

Thank you for the reply!

Yes I recently build the conda environments for snakePipes (~2 weeks ago). 

I first ran the createIndices pipeline for a human_lambda hybrid genome. At first I thought this `rchr1 1` was caused by running `bwameth` so recently. However, when I checked the [pre-made indices](https://zenodo.org/record/4471116) provided in zenodo,  the `genome.fa.bwameth.c2t` also started with `rchr1 1`. I have not yet tried running WGBS with the pre-made indices yet.

I also saw this issue running `bwameth.py` mentioned in [bwa-meth](https://github.com/brentp/bwa-meth/issues/86), but I think it has been solved. 

Lastly, I thought the `unrecognised reference name`  is what made WGBS pipeline 'stucked'. However, when I subset my original data to 1000 reads, this process can actually proceed, even though I still got `unrecognised reference name` error in the log file. 

PS: when running the subsetted data, I encountered an error in produceReport and I reported it yesterday #938 .

Thanks a lot!
Hanrong

**Comment by katsikora** [comment]
Hi Hanrong,

if your pipeline was getting stuck with the full read set but is running with a subset of reads, you might need to assign more memory to the jobs. Are you running on a cluster or with `--local` ?

Best wishes,

Katarzyna

**Comment by hanrong498** [comment]
Hi Katarzyna,

I am running it on one compute node on the cluster. Indeed in the first run, when I used the full read set, I did not assign enough memory. I will try again with this configuration: 

```
#SBATCH --nodes 1
#SBATCH --ntasks 60
#SBATCH --mem 400G
```

In this node, there are ~7GB ram per core. Then in the pipeline, I will set `--local -j 60`. I hope this could finish the run :)

But as I am new to WGBS analysis, I wonder if this `unrecognised reference name` error, even though will not kill the process, can cause any downstream analysis problem?

Thanks & best regards!
Hanrong

**Comment by hanrong498** [comment]
Hi Katarzyna,

With this new mem allocation I can finish the bwa mem. 


Additionally, I will put here a small issue I had in sambamba-markdup: 
```
sambamba-markdup: Cannot open file `/tmp//snakepipes.QZfyXB38ey/211029_1_FZ_WGBS_HOIK_p35_rep1_S1_L001/sambamba-pid1654661-markdup-jlul/PairedEndsInfoftfn19' in mode `w+' (Too many open files)
```
This can be solved by changing `ulimit -n <large number>` as suggested [here](https://github.com/biod/sambamba/issues/229). 



However I encountered the same problem `! invalid 'times' argument` as #898 . Specifically, I have the same PCA error: 

```
Quitting from lines 197-212 [unnamed-chunk-10] (tmpxprjy2ka.WGBS_QC_report_template.Rmd)
Error in `rep()`:
! invalid 'times' argument
Backtrace:
 1. FactoMineR::PCA(t(methyl[IDX, ]), graph = FALSE)
 2. FactoMineR::svd.triplet(X, row.w = row.w, col.w = col.w, ncp = ncp)
 4. base::crossprod(rep(1, nrow(V)), as.matrix(V))
```

I am running on one sample, understandably I have the same problem. In the previous issue, you mentioned you add some code to skip this PCA plot. Is this implemented as an option in the snakePipes now or that I would need to change it myself?

Thank you very much for your help!
Hanrong

**Comment by katsikora** [comment]
> unrecognised reference name

Hi Hanrong,

what does the header of the bam file coming out of the mapping look like regarding the reference sequences ?
```samtools view -H $bam | grep @SQ```

Best wishes,

Katarzyna

**Comment by katsikora** [comment]
> Hi Katarzyna,
> 
> With this new mem allocation I can finish the bwa mem.
> 
> Additionally, I will put here a small issue I had in sambamba-markdup:
> 
> ```
> sambamba-markdup: Cannot open file `/tmp//snakepipes.QZfyXB38ey/211029_1_FZ_WGBS_HOIK_p35_rep1_S1_L001/sambamba-pid1654661-markdup-jlul/PairedEndsInfoftfn19' in mode `w+' (Too many open files)
> ```
> 
> This can be solved by changing `ulimit -n <large number>` as suggested [here](https://github.com/biod/sambamba/issues/229).
> 
> However I encountered the same problem `! invalid 'times' argument` as #898 . Specifically, I have the same PCA error:
> 
> ```
> Quitting from lines 197-212 [unnamed-chunk-10] (tmpxprjy2ka.WGBS_QC_report_template.Rmd)
> Error in `rep()`:
> ! invalid 'times' argument
> Backtrace:
>  1. FactoMineR::PCA(t(methyl[IDX, ]), graph = FALSE)
>  2. FactoMineR::svd.triplet(X, row.w = row.w, col.w = col.w, ncp = ncp)
>  4. base::crossprod(rep(1, nrow(V)), as.matrix(V))
> ```
> 
> I am running on one sample, understandably I have the same issue. In the previous issue, you mentioned you add some code to skip this PCA plot. Is this implemented as an option in the snakePipes now or that I would need to change it myself?
> 
> And a related question: the next step I would indeed analyse multiple samples. My current folder structure is:
> 
> ```
> /path/to/sample1/sample1_R1.fastq.gz, sample1_R2.fastq.gz, 
> /path/to/sample2/sample2_R1.fastq.gz, sample2_R2.fastq.gz
> ```
> 
> To dump all samples to the pipeline, I would need to put all fastq.gz files into one folder? If so, can symlink be enough?
> 
> Thank you very much for your help! Hanrong

Hi Hanrong,

it looks like you're mapping large amounts of data, and sambamba chunked your large file into many small ones for marking duplicates. Then you hit the number of open files allowed by your system admin.
You can alleviate this issue by increasing the memory assigned to rule `markDupes` .

The other issue you mention - `invalid 'times' argument` is because you're running only one sample, and PCA cannot be computed. The fix will be available in the next release.
Can you edit this temporary file `tmpxprjy2ka.WGBS_QC_report_template.Rmd` and add `eval=FALSE` to the PCA Rmd section?
It now reads:
```{r warning=FALSE, message=FALSE}```
If you change it to:
```{r warning=FALSE, message=FALSE,eval=FALSE}```
the code chunk won't be executed and the report creation should run through.

Indeed, if you would like to process multiple files at the same time, it's best to symlink them into one input folder.
e.g. ```ln -t $my_folder_with_symlinks -s $file1 $file2 $file3 ...```

Best wishes,

Katarzyna

**Comment by hanrong498** [comment]
> > unrecognised reference name
> 
> Hi Hanrong,
> 
> what does the header of the bam file coming out of the mapping look like regarding the reference sequences ? `samtools view -H $bam | grep @SQ`
> 
> Best wishes,
> 
> Katarzyna

Hi Katarzyna,

This actually returns nothing.. 

```
[hhu@jed Sambamba]$ samtools view 211029_1_FZ_WGBS_HOIK_p35_rep1_S1_L001.markdup.bam | grep @SQ
[hhu@jed Sambamba]$ 
```

Best,
Hanrong

**Comment by hanrong498** [comment]
> The other issue you mention - `invalid 'times' argument` is because you're running only one sample, and PCA cannot be computed. The fix will be available in the next release. Can you edit this temporary file `tmpxprjy2ka.WGBS_QC_report_template.Rmd` and add `eval=FALSE` to the PCA Rmd section? It now reads: `{r warning=FALSE, message=FALSE}` If you change it to: `{r warning=FALSE, message=FALSE,eval=FALSE}` the code chunk won't be executed and the report creation should run through.


Hi Katarzyna,

Thank you for the tip! And sorry for my late reply regarding this issue. I cannot find this `tmpxprjy2ka.WGBS_QC_report_template.Rmd` file.

I tried here `/scratch/hhu/20220330_P1886_FIDES_DNA_METHYLATION/merged_processed/.snakemake/scripts/tmpxprjy2ka.WGBS_QC_report_template.Rmd` but this `/scripts/` folder is empty. Maybe I can change from the source code? 

Thanks! 
Hanrong

**Comment by katsikora** [comment]
Hi Hanrong,

sure, you can also  modify the source file. The updated [Rmd](https://github.com/maxplanck-ie/snakepipes/blob/develop/snakePipes/shared/rscripts/WGBS_QC_report_template.Rmd) is now available on the `develop` branch of snakePipes .

You should be able to find the "source" Rmd under:
```
$github_repo_dir/snakePipes/shared/rscripts/WGBS_QC_report_template.Rmd
``` 
and replace it e.g. with the one from the develop branch.


Hope this helps,
Best,

Katarzyna

**Comment by katsikora** [comment]
This should now be fixed in snakePipes 2.8.0 and 2.8.1.
